benchmarks/polys.py

```diff
@@ -1,17 +1,137 @@
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _SparseGCDHighDegree:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = 1 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)])
+        f = D * (-2 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)]))
+        g = D * (2 + x ** (n + 1) + sum([y[i] ** (n + 1) for i in range(n)]))
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _QuadraticNonMonicGCD:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = 1 + x ** 2 * y[0] ** 2 + sum([y[i] ** 2 for i in range(1, n)])
+        f = D * (-1 + x ** 2 - y[0] ** 2 + sum([y[i] ** 2 for i in range(1, n)]))
+        g = D * (2 + x * y[0] + sum(y[1:n])) ** 2
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _SparseNonMonicQuadratic:
+    def setup_polys(self, n):
+        self.x, *self.y = symbols("x, y1:9")
+        self.R1 = [self.x] + list(self.y)
+        self.R = ZZ[self.R1]
+        x, y, R = self.x, self.y, self.R
+        D = -1 + x * prod(y[:n])
+        f = D * (3 + x * prod(y[:n]))
+        g = D * (-3 + x * prod(y[:n]))
+        fp, gp = Poly(f, x, *y[:n]), Poly(g, x, *y[:n])
+        fpe, gpe = R.from_sympy(f), R.from_sympy(g)
+        return f, g, fp, gp, fpe, gpe, x, y, R
+
+
+class _TimePREM:
+    """Benchmarks for `prem` method in polynomials."""
+
+    def setup(self, n, method):
+        (f, g, fp, gp, fpe, gpe, x, y, R) = self.setup_polys(n)
+
+        self.values = {}
+
+        prem_f_g_x = rem(f * LC(g, x) ** (degree(f, x) - degree(g, x) + 1), g, x)
+
+        self.ref_expr = prem_f_g_x
+        self.ref_ring = R(prem_f_g_x.as_expr())
+
+        if method == 'prem':
+            self.func = lambda: prem(f, g, x)
+
+        elif method == 'poly':
+            self.func = lambda: fp.prem(gp)
+
+        elif method == 'sparse':
+            self.func = lambda: fpe.prem(gpe)
+
+    def teardown(self, n, method):
+        for key, val in self.values.items():
+            if key == 'sparse':
+                if (self.ref_ring - val) != 0:
+                    raise ValueError("Incorrect result, invalid timing:"
+                                        " %s != %s" % (self.ref_ring, val))
+
+            elif (self.ref_expr - val).simplify() != 0:
+                raise ValueError("Incorrect result, invalid timing:"
+                                    " %s != %s" % (self.ref_expr, val))
```

Comment:
The setup method will be called once for each value of `method`. The teardown method will be called also once for each value. Here the teardown method checks *all* values but it should just check the one that was just tested.

It would be better if the setup method just stored the expected result for the given format like:
```
        if method == 'prem':
            self.func = lambda: prem(f, g, x)
            self.expected = prem_f_g_x
```
Then the teardown method can just do:
```
assert self.result == self.expected
```
Also it should not be necessary to use simplify here.